source/extensions/filters/http/on_demand/on_demand_update.cc

```diff
@@ -28,6 +64,57 @@ Http::FilterHeadersStatus OnDemandRouteUpdate::decodeHeaders(Http::RequestHeader
   return filter_iteration_state_;
 }
 
+// Passed route pointer here is not null.
+void OnDemandRouteUpdate::handleOnDemandCDS(const Router::RouteConstSharedPtr& route) {
+  auto config = getConfig(route);
+  if (config == nullptr) {
+    filter_iteration_state_ = Http::FilterHeadersStatus::Continue;
+    return;
+  }
+  auto odcds = config->odcds();
+  if (odcds == nullptr) {
+    // No ODCDS? It means that on-demand discovery is disabled.
+    filter_iteration_state_ = Http::FilterHeadersStatus::Continue;
+    return;
+  }
+  if (callbacks_->clusterInfo() != nullptr) {
+    // Cluster already exists, so nothing to do here.
+    filter_iteration_state_ = Http::FilterHeadersStatus::Continue;
+    return;
+  }
+  auto entry = route->routeEntry();
+  if (entry == nullptr) {
+    // No entry? Nothing we can do here.
+    filter_iteration_state_ = Http::FilterHeadersStatus::Continue;
+    return;
+  }
+  auto cluster_name = entry->clusterName();
+  if (cluster_name.empty()) {
+    // Empty cluster name may be a result of a missing HTTP header
+    // used for getting the cluster name. Nothing we can do here.
+    filter_iteration_state_ = Http::FilterHeadersStatus::Continue;
+    return;
+  }
+  filter_iteration_state_ = Http::FilterHeadersStatus::StopIteration;
+  auto callback = std::make_unique<Upstream::ClusterDiscoveryCallback>(
+      [this](Upstream::ClusterDiscoveryStatus cluster_status) {
+        onClusterDiscoveryCompletion(cluster_status);
+      });
+  cluster_discovery_handle_ = odcds->requestOnDemandClusterDiscovery(
+      cluster_name, std::move(callback), std::chrono::milliseconds(5000));
```

Comment:
Suggest making this a constant. And/or configurable. This something that should be coordinated with what @lambdai is doing for TCP proxy.

Reply:
Yeah, I think I'll add the timeout field to the OnDemand message, just like in tcp_proxy.

Reply:
Added the timeout field.